src/number_theory/fermat4.lean

```diff
@@ -0,0 +1,322 @@
+  { apply ne_zero_pow (dec_trivial : 2 ≠ 0),
+    rw hs, apply mul_ne_zero, { exact ne_of_gt h4}, { exact hrsz } },
+  obtain ⟨i, hi⟩ := int.sqr_of_coprime hcp hs.symm,
+  -- use m is positive to exclude m = - i ^ 2
+  have hi' : ¬ m = - i ^ 2,
+  { by_contradiction h1,
+    have hit : - i ^ 2 ≤ 0, apply neg_nonpos.mpr (pow_two_nonneg i),
+    rw ← h1 at hit,
+    apply absurd h4 (not_lt.mpr hit) },
+  replace hi : m = i ^ 2, { apply or.resolve_right hi hi' },
+  have hi0 : i ≠ 0,
+  { intro h0, rw [h0, zero_pow (dec_trivial : 0 < 2)] at hi, apply absurd hi (ne_of_gt h4) },
+  rw mul_comm at hs,
+  rw [int.gcd_comm] at hcp,
+  -- obtain d such that r * s = d ^ 2
+  obtain ⟨d, hd⟩ := int.sqr_of_coprime hcp hs.symm,
+  -- (b / 2) ^ 2 and m are positive so r * s is positive
+  have hd' : ¬ r * s = - d ^ 2,
+  { by_contradiction h1,
+    rw h1 at hs,
+    have h2 : (b / 2) ^ 2 ≤ 0,
+    { rw [hs, (by ring : - d ^ 2 * m = - (d ^ 2 * m))], apply neg_nonpos.mpr,
+      apply (zero_le_mul_right h4).mpr (pow_two_nonneg d) },
+    have h2' : 0 ≤ (b / 2) ^ 2, { apply pow_two_nonneg (b / 2) },
+    exact absurd (lt_of_le_of_ne h2' (ne.symm (pow_ne_zero _ h2b0))) (not_lt.mpr h2) },
+  replace hd : r * s = d ^ 2, { apply or.resolve_right hd hd' },
+  -- r = +/- j ^ 2
+  obtain ⟨j, hj⟩ := int.sqr_of_coprime htt4 hd,
+  have hj0 : j ≠ 0,
+  { intro h0, rw [h0, zero_pow (dec_trivial : 0 < 2), neg_zero, or_self] at hj,
+    apply left_ne_zero_of_mul hrsz hj },
+  rw mul_comm at hd,
+  rw [int.gcd_comm] at htt4,
+  -- s = +/- k ^ 2
+  obtain ⟨k, hk⟩ := int.sqr_of_coprime htt4 hd,
+  have hk0 : k ≠ 0,
+  { intro h0, rw [h0, zero_pow (dec_trivial : 0 < 2), neg_zero, or_self] at hk,
+    apply right_ne_zero_of_mul hrsz hk },
+  have hj2 : r ^ 2 = j ^ 4, { cases hj with hjp hjp; { rw hjp, ring } },
+  have hk2 : s ^ 2 = k ^ 4, { cases hk with hkp hkp; { rw hkp, ring } },
+  -- from m = r ^ 2 + s ^ 2 we now get a new solution to a ^ 4 + b ^ 4 = c ^ 2:
+  have hh : i ^ 2 = j ^ 4 + k ^ 4, { rw [← hi, htt3, hj2, hk2] },
+  have hn : n ≠ 0, { rw ht2 at hb20, apply right_ne_zero_of_mul hb20 },
+  -- and it has a smaller c: from c = m ^ 2 + n ^ 2 we see that m is smaller than c, and i ^ 2 = m.
+  have hic : int.nat_abs i < int.nat_abs c,
+  { apply int.coe_nat_lt.mp, rw ← (int.eq_nat_abs_of_zero_le (le_of_lt hc)),
+    apply gt_of_gt_of_ge _ (int.abs_le_mul_self i),
+    rw [← hi, ht3],
+    apply gt_of_gt_of_ge _ (int.le_mul_self m),
+    exact lt_add_of_pos_right (m ^ 2) (pow_two_pos_of_ne_zero n hn) },
+  have hic' : int.nat_abs c ≤ int.nat_abs i,
+  { apply (h.2 j k i),
+    exact and.intro hj0 (and.intro hk0 (and.intro hi0 hh.symm)) },
+  apply absurd (not_le_of_lt hic) (not_not.mpr hic')
+end
+
+lemma not_fermat_42 {a b c : ℤ} : ¬fermat_42 a b c :=
+begin
+  intro h,
+  obtain ⟨a0, b0, c0, ⟨hf, h2, hp⟩⟩ := exists_pos_odd_minimal_fermat_42 h,
+  apply not_minimal_fermat_42 hf h2 hp
+end
+
+lemma not_fermat_4 {a b c : ℤ} (ha : a ≠ 0) (hb : b ≠ 0) (hc : c ≠ 0) :
```

Comment:
You shouldn't need `hc`, right?

Reply:
Yep, I missed that. I've removed it throughout and also did some refactoring using a fermat_42 namespace. Sorry, should probably not have done this after submitting the PR, but it suddenly seemed to make much more sense to have things named e.g. `fermat42.not_minimal` instead of `not_minimal_fermat_42`.